src/torch_ucc.cpp

```diff
@@ -718,6 +737,88 @@ ProcessGroupUCC::~ProcessGroupUCC() {
   }
 }
 
+void ProcessGroupUCC::runHealthCheck() {
+  // Run health check in a separate thread and wait on CV to handle timeouts.
+  // This design allows us to handle hangs.
+
+  // When size_ is 1, there is no need to do any communication at all.
+  if (size_ == 1) return;
+
+  struct HealthCheckData {
+    std::mutex healthCheckMutex;
+    std::condition_variable healthCheckCv;
+    bool ucxHealthCheckSuccess = false;
+    bool uccHealthCheckSuccess = false;
+    std::exception_ptr healthCheckException;
+  } healthCheckData;
+  auto t = std::thread([&healthCheckData, this]() {
+    try {
+      auto oob = std::make_shared<torch_ucc_oob_coll_info_t>();
+      oob->rank = this->oob->rank;
+      oob->size = this->oob->size;
+      oob->store = this->oob->store;
+
+      std::vector<ucp_ep_h> eps;
+      ucc_team_h team = nullptr;
+      uint32_t comm_id;
+
+      // auto comm = std::make_unique<CommPG>(logger, oob, c10::kCPU, true);//comm_id, c10::kCPU, oob, logger, true);
+      auto comm = CommPG::get_comm(comm_id, c10::kCPU, oob, logger, true);
+      comm->ucx_connect_eps(eps, oob);
+      comm->ucx_disconnect_eps(eps, oob);
+      TORCH_UCC_LOG_INFO(TORCH_UCC_HEALTH_CHECK, "UCX library health check succeed.");
+      // Mark ucx health check as complete.
+      {
+        std::lock_guard<std::mutex> lk(healthCheckData.healthCheckMutex);
+        healthCheckData.ucxHealthCheckSuccess = true;
+      }
+
+      comm->ucc_create_team(team, oob);
+      comm->ucc_destroy_team(team);
+      TORCH_UCC_LOG_INFO(TORCH_UCC_HEALTH_CHECK, "UCC library health check succeed.");
+      // Mark ucc health check as complete.
+      {
+        std::lock_guard<std::mutex> lk(healthCheckData.healthCheckMutex);
+        healthCheckData.uccHealthCheckSuccess = true;
+      }
+
+      comm = nullptr;
+      oob = nullptr;
+      // Notify main thread the health check is complete.
+      healthCheckData.healthCheckCv.notify_one();
+    } catch (const std::exception& e) {
+      // Populate exception ptr.
+      healthCheckData.healthCheckException = std::current_exception();
+      // Unblock waiting main thread which will report exception.
+      healthCheckData.healthCheckCv.notify_one();
+    } // Unknown exceptions will just cause the program to terminate.
+  });
+  // We don't need to join the thread, just need to verify health check via the
+  // CV. Hence we detach the thread here.
+  t.detach(); // NOLINT
+  LOG(INFO) << "[Rank " << rank_ << "]"
+            << " will wait up to " << timeout_.count()
+            << " msec for NCCL health check to complete.";
```

Comment:
use TORCH_UCC logging to print unified log info, e.g.,

```suggestion
  TORCH_UCC_LOG_INFO(
      TORCH_UCC_HEALTH_CHECK,
      c10::str(
          "will wait up to ", timeout_.count(),
          " msec for UCC health check to complete.")
  );
```

Reply:
fixed